components/org.wso2.carbon.identity.auth.service/src/main/java/org/wso2/carbon/identity/auth/service/handler/AuthenticationHandler.java

```diff
@@ -107,76 +105,64 @@ protected void postAuthenticate(MessageContext messageContext, AuthenticationRes
+                    String username = resolveUserNameForOrganizationUser((AuthenticatedUser) user);
+                    PrivilegedCarbonContext.getThreadLocalCarbonContext().setUsername(username);
+                }
+
+                // Set the user's resident organization if user is accessing an organization
+                if (user instanceof AuthenticatedUser &&
+                        StringUtils.isNotEmpty(((AuthenticatedUser) user).getAccessingOrganization())) {
                     PrivilegedCarbonContext.getThreadLocalCarbonContext()
-                            .setUserResidentOrganizationId(userResidentOrganization);
-                    if (((AuthenticatedUser) user).isFederatedUser()) {
-                        updateUserNameInContextForOrganizationSsoUsers(userResidentOrganization);
-                    }
+                            .setUserResidentOrganizationId(((AuthenticatedUser) user).getUserResidentOrganization());
                 }
             }
         }
     }
 
-    private void updateUserNameInContextForOrganizationSsoUsers(String userResidentOrganization) {
+    private String resolveUserNameForOrganizationUser(AuthenticatedUser authenticatedUser) {
 
         try {
             String tenantDomain = AuthenticationServiceHolder.getInstance().getOrganizationManager()
-                    .resolveTenantDomain(userResidentOrganization);
+                    .resolveTenantDomain(authenticatedUser.getUserResidentOrganization());
             int tenantId = AuthenticationServiceHolder.getInstance().getRealmService().getTenantManager()
                     .getTenantId(tenantDomain);
-            RealmService realmService = AuthenticationServiceHolder.getInstance().getRealmService();
-            UserRealm tenantUserRealm = realmService.getTenantUserRealm(tenantId);
-            if (tenantUserRealm != null) {
-                String userId = PrivilegedCarbonContext.getThreadLocalCarbonContext().getUserId();
-                org.wso2.carbon.user.core.common.User user =
-                        ((AbstractUserStoreManager) tenantUserRealm.getUserStoreManager()).getUser(userId, null);
-                if (user != null && StringUtils.isNotEmpty(user.getUsername())) {
-                    PrivilegedCarbonContext.getThreadLocalCarbonContext()
-                            .setUsername(user.getDomainQualifiedUsername());
-                }
-            }
-        } catch (OrganizationManagementException | UserStoreException e) {
+            AbstractUserStoreManager userStoreManager = getAbstractUserStoreManager(tenantId);
+            return userStoreManager.getUserNameFromUserID(authenticatedUser.getUserId());
+        } catch (OrganizationManagementException | UserStoreException | UserIdNotFoundException e) {
             LOG.error("Authenticated user's username could not be resolved.", e);
         }
+        return StringUtils.EMPTY;
```

Comment:
Handle from the invocation.

Reply:
In the previous implementation, when username didn't get resolved, it simply skip setting the username in the context. In that cases, username might have been set with user-id which is wrong.

Here if fails to resolve the username, it set the username as empty, so that if any downstream task requires to fetch the username, it will not relay on the username in the context as it is set to empty, hence might rely on the user-id if it is properly set in the context.

As setting username, user-id in the context are not directly related to post authentication, but added for any usecase where the request initiator is required (ex; resource creation, auditing etc.), only added an error log instead of throwing an exception.